### PR TITLE
try/catch apply

### PR DESCRIPTION
**Pull request by sbmaxx**
Fixes for https://github.com/bem/bem-xjst/issues/113

### CODE REVIEW

**Comment by indutny** [comment]
I think it is very dangerous change. It may obscure things running in production, without giving our any kind of introspection. IMO, we should find a better way for it.

**Comment by sbmaxx** [comment]
@indutny any ideas?

**Comment by indutny** [comment]
I think we should save the error here: https://github.com/bem/bem-xjst/blob/master/lib/bemhtml/runtime/match.js#L211 and re-throw it after restoring all values: https://github.com/bem/bem-xjst/blob/master/lib/bemhtml/runtime/match.js#L217

**Comment by sbmaxx** [comment]
@indutny like this https://github.com/bem/bem-xjst/commit/5b10fb3fa4da600ece866db324d0e5a6cf1aec98 ?

**Comment by indutny** [comment]
@sbmaxx this is not as great either. Why not wrap `template.body.call()` into try-catch in-place and store exception to the local variable there. If that local variable is non-empty at the end of the `exec` - throw exception.

Does it sound right?

**Comment by sbmaxx** [comment]
Just because of perf:

**Inined** `try/catch`:
```
sbmaxx@sbmaxx-mba ~/Develop/bem-xjst/bench $ make -B -C ../ && node run --compare
make: Circular lib/bemhtml/runtime/bundle.js <- lib/bemhtml/runtime/bundle.js dependency dropped.
./node_modules/.bin/browserify --standalone BEMHTML lib/bemhtml/runtime/index.js -o lib/bemhtml/runtime/bundle.js
render:basic:next x 240,096 ops/sec ±1.05% (86 runs sampled)
render:basic:prev x 299,737 ops/sec ±1.22% (85 runs sampled)
render:islands:next x 473 ops/sec ±1.34% (84 runs sampled)
render:islands:prev x 617 ops/sec ±1.38% (84 runs sampled)
render:showcase:next x 999 ops/sec ±1.28% (83 runs sampled)
render:showcase:prev x 1,212 ops/sec ±2.72% (83 runs sampled)
# one more try
render:basic:next x 244,273 ops/sec ±1.11% (86 runs sampled)
render:basic:prev x 296,970 ops/sec ±1.06% (87 runs sampled)
render:islands:next x 469 ops/sec ±1.54% (84 runs sampled)
render:islands:prev x 590 ops/sec ±2.09% (84 runs sampled)
render:showcase:next x 819 ops/sec ±5.02% (82 runs sampled)
render:showcase:prev x 1,094 ops/sec ±2.94% (81 runs sampled)
```

**Prototype.tryCatch**:
```
sbmaxx@sbmaxx-mba ~/Develop/bem-xjst/bench $ make -B -C ../ && node run --compare
make: Circular lib/bemhtml/runtime/bundle.js <- lib/bemhtml/runtime/bundle.js dependency dropped.
./node_modules/.bin/browserify --standalone BEMHTML lib/bemhtml/runtime/index.js -o lib/bemhtml/runtime/bundle.js
render:basic:next x 292,620 ops/sec ±1.11% (87 runs sampled)
render:basic:prev x 302,492 ops/sec ±1.15% (86 runs sampled)
render:islands:next x 554 ops/sec ±1.56% (85 runs sampled)
render:islands:prev x 572 ops/sec ±2.22% (85 runs sampled)
render:showcase:next x 1,051 ops/sec ±2.10% (85 runs sampled)
render:showcase:prev x 1,166 ops/sec ±2.25% (82 runs sampled)
```

**Comment by sbmaxx** [comment]
@indutny seems like inlined try/catch has bigger performance penalty, can you check this?

**Comment by indutny** [comment]
@sbmaxx I will definitely do, but later today.

**Comment by sbmaxx** [comment]
ok, thx! :)

**Comment by indutny** [comment]
Btw, I have strong doubts that this thing is going to work properly even after this fix. Throwing exceptions will leave dangling changes made by `local(...)`. Perhaps we are trying to fix the thing that should not be working by-design? cc @veged

**Comment by veged** [comment]
let's discuss it offline at our regular meeting...

**Comment by sbmaxx** [comment]
Any news?

**Comment by zxqfox** [comment]
> seems like inlined try/catch has bigger performance penalty, can you check this?

@sbmaxx Sure it is. Any code block with try-catch can't be optimized by v8. Why not to make universal `tryCatch` and use it everywhere in project?

```
function tryCatch(fn, cb) {
  try { return fn(); } catch(e) { cb(e); }
}
//...
  var err;
  out = tryCatch(template.body.bind(context), function (e) { err = e; });
// ...
if (err) throw err;
return out;
```

File: lib/bemhtml/runtime/match.js [hunk 1]
```diff
@@ -208,7 +216,7 @@ Match.prototype.exec = function exec(context) {
 
   var out;
   if (typeof template.body === 'function')
-    out = template.body.call(context);
+    out = this.tryCatch(template.body,  context);
```
Comment (zxqfox) [inline lib/bemhtml/runtime/match.js]: double space before context

File: lib/bemhtml/runtime/match.js [hunk 2]
```diff
@@ -166,6 +166,14 @@ Match.prototype.push = function push(template) {
   this.maskSize = this.mask.length;
 };
 
+Match.prototype.tryCatch = function(fn, ctx) {
+  try {
+    return fn.call(ctx);
+  } catch (e) {
+    return e;
```
Comment (indutny) [inline lib/bemhtml/runtime/match.js]: Let's create a property called `.thrownError` in constructor, and initialize it to `null`.

Comment (indutny) [inline lib/bemhtml/runtime/match.js]: Instead of `return e`, we should now do `this.thrownError = e`.

File: lib/bemhtml/runtime/match.js [hunk 3]
```diff
@@ -166,6 +166,14 @@ Match.prototype.push = function push(template) {
   this.maskSize = this.mask.length;
 };
 
+Match.prototype.tryCatch = function(fn, ctx) {
+  try {
+    return fn.call(ctx);
```
Comment (indutny) [inline lib/bemhtml/runtime/match.js]: And before the call it should be set to `null` again.

File: lib/bemhtml/runtime/match.js [hunk 4]
```diff
@@ -216,6 +224,9 @@ Match.prototype.exec = function exec(context) {
   this.bemhtml.match = oldMatch;
   this.restoreDepth(save);
 
+  if (out instanceof Error)
```
Comment (indutny) [inline lib/bemhtml/runtime/match.js]: This should check if `.thrownError` is not `null`, load it, cleanup the field (set to `null`), and throw  it.

File: lib/bemhtml/runtime/match.js (hunk 1 above)
Comment (sbmaxx) [inline lib/bemhtml/runtime/match.js]: :gem: eyes ;)

i think we need jscs here

Comment (indutny) [inline lib/bemhtml/runtime/match.js]: We have it.

**Comment by sbmaxx** [comment]
@indutny fixed. Also added some tests about this fix

cc @zxqfox, @veged

File: lib/bemhtml/runtime/match.js [hunk 5]
```diff
@@ -216,7 +227,15 @@ Match.prototype.exec = function exec(context) {
...
 
+  var e;
+  if (this.thrownError !== null) {
+    e = this.thrownError;
+    this.thrownError = null;
+    throw e;
+  }
+
   return out;
+
```
Comment (zxqfox) [inline lib/bemhtml/runtime/match.js]: Quitely fine PR imho, except this line. ;-)

File: lib/bemhtml/runtime/match.js [hunk 6]
```diff
@@ -216,7 +227,15 @@ Match.prototype.exec = function exec(context) {
   this.bemhtml.match = oldMatch;
   this.restoreDepth(save);
 
+  var e;
+  if (this.thrownError !== null) {
+    e = this.thrownError;
```
Comment (zxqfox) [inline lib/bemhtml/runtime/match.js]: NonBlocker: btw, you can reduce 1 line of code with moving this to `var`. And also you can replace `this.thrown` in `if` with just `e`.

Comment (indutny) [inline lib/bemhtml/runtime/match.js]: Yeah, please move `var e` here.

File: lib/bemhtml/runtime/match.js (hunk 5 above)
Comment (indutny) [inline lib/bemhtml/runtime/match.js]: Agree :)

File: test/runtime-test.js [hunk 7]
```diff
@@ -880,4 +880,44 @@ describe('BEMHTML compiler/Runtime', function() {
...
+    it('should throw errors with many applies', function() {
+      assert.throws(function() {
+        template.apply(bemjson);
+      });
+      assert.throws(function() {
+        template.apply(bemjson);
+      });
+    });
+  });
+
```
Comment (indutny) [inline test/runtime-test.js]: Excessive newline.

**Comment by indutny** [comment]
Do benchmarks still look good?

File: lib/bemhtml/runtime/match.js (hunk 6 above)
Comment (sbmaxx) [inline lib/bemhtml/runtime/match.js]: done

**Comment by sbmaxx** [comment]
> Do benchmarks still look good?

```bash
sbmaxx@sbmaxx-mba ~/Develop/bem-xjst/bench $ node run.js --compare
render:basic:next x 176,977 ops/sec ±8.47% (78 runs sampled)
render:basic:prev x 301,332 ops/sec ±1.18% (86 runs sampled)
render:islands:next x 607 ops/sec ±1.09% (86 runs sampled)
render:islands:prev x 580 ops/sec ±1.41% (86 runs sampled)
render:showcase:next x 1,146 ops/sec ±0.71% (86 runs sampled)
render:showcase:prev x 1,178 ops/sec ±2.34% (84 runs sampled)

sbmaxx@sbmaxx-mba ~/Develop/bem-xjst/bench $ node run.js --compare
render:basic:next x 294,854 ops/sec ±1.00% (87 runs sampled)
render:basic:prev x 297,137 ops/sec ±1.05% (87 runs sampled)
render:islands:next x 522 ops/sec ±2.91% (81 runs sampled)
render:islands:prev x 596 ops/sec ±1.27% (86 runs sampled)
render:showcase:next x 1,158 ops/sec ±1.30% (85 runs sampled)
render:showcase:prev x 1,182 ops/sec ±2.63% (84 runs sampled)

sbmaxx@sbmaxx-mba ~/Develop/bem-xjst/bench $ node run.js --compare
render:basic:next x 279,900 ops/sec ±1.06% (86 runs sampled)
render:basic:prev x 310,417 ops/sec ±0.42% (88 runs sampled)
render:islands:next x 593 ops/sec ±1.42% (85 runs sampled)
render:islands:prev x 607 ops/sec ±1.61% (87 runs sampled)
render:showcase:next x 1,108 ops/sec ±1.30% (85 runs sampled)
render:showcase:prev x 1,233 ops/sec ±2.03% (85 runs sampled)
```

**Comment by indutny** [comment]
Landed with minor fixes in 1ab21c35236b861bae8ffd0d79c3ef1713982d69. Thank you!

Btw, stable benchmark results could be obtained with `--min-samples` flag:

```bash
$ node bench/run.js --min-samples=1000 --compare
render:basic:next x 877,088 ops/sec ±0.26% (1038 runs sampled)
render:basic:prev x 935,638 ops/sec ±0.19% (1038 runs sampled)
render:islands:next x 1,412 ops/sec ±0.28% (1035 runs sampled)
render:islands:prev x 1,458 ops/sec ±0.32% (1036 runs sampled)
render:showcase:next x 2,598 ops/sec ±0.34% (1035 runs sampled)
render:showcase:prev x 2,869 ops/sec ±0.25% (1038 runs sampled)
```

**Comment by zxqfox** [comment]
Great! Thanks

**Pull request closed without merging**